### PR TITLE
Update the docker build for recent changes.

### PR DESCRIPTION
**Pull request by drgrice1**
Run `npm install` from the `pg/htdocs` location.  Also both of the `npm install` commands need to be run with `--unsafe-perm` so that the `prepare` script will run.

Don't try to change ownership and permissions on the now non-existent applets directory.

Fix the bin/addcourse script for the new WWSafe.pm location.  This script was missed in #1533.

Add the Ubuntu package libyaml-libyaml-perl which is needed for the new PGEnvironment.pm module.

Note that to test this you will need to change line 81 of docker-compose.yml to `- WEBWORK2_GIT_URL=https://github.com/drgrice1/webwork2.git` and line 95 to `- WEBWORK2_BRANCH=fix-docker`.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -95,9 +95,9 @@ services:
             - WEBWORK2_BRANCH=develop
             - PG_BRANCH=develop
 
-        # If you are using the 2 stage build process uncomment the next line
+        # If you would like a 1 stage build process comment out the next line, and just run "docker-compose build".
         dockerfile: DockerfileStage2
-        # and first run
+        # For the 2 stage build process run
         #     docker build --tag webwork-base:forWW216 -f DockerfileStage1 .
```
Comment (taniwallach) [inline docker-compose.yml:101]: I would recommend the tag `webwork-base:forWW217` here and in the relevant `FROM` line in `DockerfileStage2`

**Review by taniwallach — COMMENTED** [review]
I have not yet tested this. The changes to the Docker files all look correct to me.
Will test soon.

**Review by taniwallach — APPROVED** [review]
The docker build certainly works.

When I ran `bin/addcourse` as root (the default user when running a bash shell in our Docker container) it created the course directory as being owned by root. It might be nice if the script tried to fix this if that occurred or at least issued a warning.
```
drwxr-xr-x 7 root     root       4096 Jan 31 19:48 test111
```
and that lead to permission problems such as:
```
failed to open /opt/webwork/courses/test111/logs/login.log for writing: Permission denied at /opt/webwork/webwork2/lib/WeBWorK/Utils.pm line 824.
```

**Comment by drgrice1** [comment]
> When I ran `bin/addcourse` as root (the default user when running a bash shell in our Docker container) it created the course directory as being owned by root. It might be nice if the script tried to fix this if that occurred or at least issued a warning.

Note that the usual usage of the `addcourse` script is to create the admin course when webwork is first installed.  At that time in the installation instructions the courses directory is still under the ownership of the user.  The ownership and permissions are changed later in the installation instructions.  In the docker build this is done the same way.

I don't recommend running the script as root.  You should run it as the `www-data` user (or the server user whatever that may be on your system).  You can do this with
```
sudo -u www-data ./bin/addcourse options
```

**Comment by drgrice1** [comment]
Also, note that behavior is not new.  The script has always worked that way.  Here I was just trying to get it working again with its old behavior.  Note that this is a critical issue due to the fact that this script is used in the installation process, and is currently unusable.

**Comment by pstaabp** [comment]
Trying to test this and probably this is a case where I'm missing some docker concept.   I was able to get the docker running originally, and noticed the WWSafe error, I realized I wasn't pulling from @drgrice1 branch, so changed those lines in the `docker-compose.yml` file.

Rerunning `docker-build .` results in the following error:
```
 > [base 3/4] RUN echo Cloning branch $WEBWORK2_BRANCH from $WEBWORK2_GIT_URL   && echo git clone --single-branch --branch ${WEBWORK2_BRANCH} --depth 1 $WEBWORK2_GIT_URL   && git clone --single-branch --branch ${WEBWORK2_BRANCH} --depth 1 $WEBWORK2_GIT_URL   && rm -rf webwork2/.git webwork2/{*ignore,Dockerfile,docker-compose.yml,docker-config}:
#10 0.462 Cloning branch from
#10 0.462 git clone --single-branch --branch --depth 1
#10 0.464 fatal: repository '1' does not exist
```

It doesn't appear that the `WEBWORK2_BRANCH` and `WEBWORK2_GIT_URL` arguments are not getting passed in. I changed the lines exactly like the top of this suggested.

**Comment by drgrice1** [comment]
You should not be running `docker build .`.  You should be running `docker-compose build`.

**Comment by drgrice1** [comment]
For the two stage build process you need to use
```
docker build --tag webwork-base:forWW217 -f DockerfileStage1 .
docker-compose build
```
If you want the one stage build, make sure to comment out the line `dockerfile: DockerfileStage2` in `docker-compose.yml`.

**Comment by pstaabp** [comment]
No longer are there errors on the build using either DockerFile or DockerStage1, but on `docker-compose up`, I'm getting the following:
```
Can't locate WWSafe.pm in @INC (you may need to install the WWSafe module) (@INC contains: /opt/webwork/webwork2/lib /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.30.0 /usr/local/share/perl/5.30.0 /usr/lib/x86_64-linux-gnu/perl5/5.30 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.30 /usr/share/perl/5.30 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /opt/webwork/webwork2/lib/WeBWorK/CourseEnvironment.pm line 56.
```

The `/opt/webwork/pg/lib` directory is missing from `@INC`.  It appears that is should be there from the `webwork.apache2.4-config`

**Review by pstaabp — APPROVED** [review]
Retried from scratch today and everything works now.

**Comment by taniwallach** [comment]
Regading the error
```
Can't locate WWSafe.pm in @INC (you may need to install the WWSafe module) (@INC contains: /opt/webwork/webwork2/lib /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.30.0 /usr/local/share/perl/5.30.0 /usr/lib/x86_64-linux-gnu/perl5/5.30 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.30 /usr/share/perl/5.30 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /opt/webwork/webwork2/lib/WeBWorK/CourseEnvironment.pm line 56.
```

I ran into this upgrading a server using Docker today. The problem was that I was mounting via `docker-compose.yml` a locally modified version of `/etc/apache2/conf-enabled/webwork.conf` (from a WW 2.15 system) which did not have the relevant changes needed. For those who run into trouble - It needs to be updated from the revised `webwork2/conf/webwork.apache2.4-config.dist`.

**Pull request merged**